### ISSUE TITLE
I'd like to submit a PR, but first...

### ISSUE THREAD

**Issue by housni** [gruntwork-io/bash-commons#8]
I love this repo! There's a great set of tools here for users.

I would like to submit a few PR's but I wanted to first clarify if it's something you would use.

1. Generally, using `#!/usr/bin/env bash` is considered compatible so I'd like to make that change.
2. I'd like to make all the code `shellcheck` compatible.
3. I think it would be a good idea to set things like `pipefail`, `nounset`, `errtrace`, `functrace`, etc, in the scripts. Especially `pipefail` because I can see the scripts have some pipes so if commands in the pipes fail, we are not going to see those failures unless they are obvious. See: https://www.davidpashley.com/articles/writing-robust-shell-scripts/
4. I'd like to add a lot more functionality (I'm sure you won't have any objection to this :)). I can port some of the functionality in my own Bash library (https://github.com/housni/sheldon) to this repo, making sure I stick to the guidelines in the README.

Thoughts?

**Comment by brikis98** [comment]
> I love this repo! There's a great set of tools here for users.

Thank you!

> Generally, using #!/usr/bin/env bash is considered compatible so I'd like to make that change.

Can you provide more info/context around this?

> I'd like to make all the code shellcheck compatible.

That sounds like a great idea!

> I think it would be a good idea to set things like pipefail, nounset, errtrace, functrace, etc, in the scripts. Especially pipefail because I can see the scripts have some pipes so if commands in the pipes fail, we are not going to see those failures unless they are obvious. See: https://www.davidpashley.com/articles/writing-robust-shell-scripts/

This sounds great too. I wasn't familiar with a few of these, thx!

> I can port some of the functionality in my own Bash library (https://github.com/housni/sheldon) to this repo, making sure I stick to the guidelines in the README.

Cool library! Contributions are very welcome 👍 

What were you thinking of adding?

BTW, if you're going to be following our conventions, take note of [one issue](https://github.com/gruntwork-io/bash-commons/issues/9) we recently found and still need to fix!

**Comment by housni** [comment]
> > Generally, using #!/usr/bin/env bash is considered compatible so I'd like to make that change.
> 
> Can you provide more info/context around this?
Sure. The rationale is that, `bash` doesn't always live in `/usr/bin` or `/bin` but `env` always exists at `/usr/bin/env`. Since `env` knows where `bash` lives, we can use it to get the correct `bash` interpreter for that system.
References:

1. https://stackoverflow.com/questions/21612980/why-is-usr-bin-env-bash-superior-to-bin-bash
2. https://en.wikipedia.org/wiki/Shebang_(Unix)#Portability

> What were you thinking of adding?
A bunch of things like string and array functions as I have implemented them here (but according to your coding standards :)): https://github.com/housni/sheldon/tree/master/lib/sheldon/util
I'll be creating a separate PR for the extra functions so you can decide on what works and what doesn't.

> BTW, if you're going to be following our conventions, take note of [one issue](https://github.com/gruntwork-io/bash-commons/issues/9) we recently found and still need to fix!
Noted! I did notice the existing syntax and realized I had not seen that before and I was wondering if it was one of those Bash-isms that I didn't know about :)

**Comment by brikis98** [comment]
> The rationale is that, bash doesn't always live in /usr/bin or /bin but env always exists at /usr/bin/env. Since env knows where bash lives, we can use it to get the correct bash interpreter for that system.

Thanks for the explanation. This sounds good to me!

> A bunch of things like string and array functions as I have implemented them here (but according to your coding standards :)): https://github.com/housni/sheldon/tree/master/lib/sheldon/util
I'll be creating a separate PR for the extra functions so you can decide on what works and what doesn't.

That sounds good too!

**Comment by brikis98** [comment]
Perhaps these should go into a `flags.sh` that all the scripts `source` right at the top?

**Comment by brikis98** [comment]
A `bootstrap.sh` sounds good. Note that we'd need to update the docs to recommend `source`ing this file right at the top and update the tests in this repo to do the same.

**Comment by housni** [comment]
Sounds great!

Also, I'll have CircleCI run `shellcheck` as a part of CI.

**Comment by housni** [comment]
I think that PR is ready to be merged in. Can we hook up CircleCI? I run `shellcheck` via CircleCI now.
Btw, added `bootstrap.sh` caused the tests to exit. There's a lot of tests so I thought it would be best if it had it's own PR in the interest of completing PR #11.

**Comment by housni** [comment]
Scratch that, I have one more thing to do before we can merge!

**Comment by brikis98** [comment]
> Scratch that, I have one more thing to do before we can merge!

OK, let me know when it's ready for another review!

**Comment by housni** [comment]
Ok, I think this PR is ready for a review now.
Just one thing: https://github.com/gruntwork-io/bash-commons/pull/11#pullrequestreview-173726273

**Comment by housni** [comment]
Btw, is CircleCI hooked up? It should run `shellcheck` and `bats`.

**Comment by brikis98** [comment]
Yes, CircleCI is hooked up, but does not run on external PRs.

**Comment by housni** [comment]
Btw, the PR is ready to go, whenever you want to merge it.

**Comment by brikis98** [comment]
My bad, I lost track of this one. Thank you for the reminder and the great contribution. Merged now! I'll mark this issue as closed.

**Comment by housni** [comment]
Awesome! Glad to be able to contribute!
